File: Sources/TintCore/Palettes/Palette.swift
/// One palette colour and the share of the image it covers (0...1).
public struct Swatch: Hashable, Sendable {
    public var color: Rgb
    public var share: Double

    public init(color: Rgb, share: Double) {
        self.color = color
        self.share = share
    }

    public var lab: Lab { color.lab }
}

/// The colours extracted from an image, most common first.
public struct Palette: Hashable, Sendable {
    public var swatches: [Swatch]

    public init(swatches: [Swatch]) { self.swatches = swatches }

    /// Average lightness weighted by coverage, 0...100. Below ~50 reads as a dark image.
    public var lightness: Double { swatches.reduce(0) { $0 + $1.lab.l * $1.share } }

    public var isDark: Bool { lightness < 50 }
}

File: src/Tint.Core/Palettes/Palette.cs
using Tint.Core.Colors;

namespace Tint.Core.Palettes;

/// <summary>One palette colour and the share of the image it covers (0..1).</summary>
public sealed record Swatch(Rgb Color, double Share)
{
    public Lab Lab => Color.ToLab();
}

/// <summary>The colours extracted from an image, most common first.</summary>
public sealed record Palette(IReadOnlyList<Swatch> Swatches)
{
    /// <summary>Average lightness weighted by coverage, 0..100. Below ~50 reads as a dark image.</summary>
    public double Lightness => Swatches.Sum(s => s.Lab.L * s.Share);

    public bool IsDark => Lightness < 50;
}
